[SPARK-38796][SQL] Update documentation for number format strings with the {try_}to_number functions (#36950)
apache/spark: docs/sql-ref-number-pattern.md

  limitations under the License.
---

### Description

Functions such as `to_number` and `try_to_number` support converting between values of string and
Decimal type. Such functions accept format strings indicating how to map between these types.

> ### Syntax

Reviewer: Shall we have the highlights for all the functions?
E.g.
```
to_number(string, format);
try_to_number(string, format);
```

Author: Sure, this sounds good. Added this to the docs.

  A leading `9` indicates that these digits are optional.

  The input value must not be larger than the number of digits to the left of the decimal point 
  allowed by the format string.

  Digits to the right of the decimal point in the format string indicate the most digits that the
  input value may have to the right of the decimal point.

- **`.`** or **`D`**

  Specifies the position of the decimal point.

  The input value does not need to include a decimal point.

- **`,`** or **`G`**

  Specifies the position of the `,` grouping (thousands) separator.

  There must be a `0` or `9` to the left and right of each grouping separator. 

- **`$`**

  Specifies the location of the `$` currency sign. This character may only be specified once.

- **`S`** 

  Specifies the position of an optional '+' or '-' sign. This character may only be specified once.

- **`MI`**

  Specifies an optional `-` sign at the end, but no `+`.

- **`PR`**

  Maps negative input values to wrapping angle brackets (`<1>`) in the corresponding string.

  Positive input values do not receive wrapping angle brackets.

### Examples

> The following examples use the `try_to_number` SQL function which accepts an input string as the

Reviewer: It would be great if we can have examples for to_number as well

Author: Added this to the examples section at the end.

  limitations under the License.
---

### Description

> Functions such as `to_number` and `try_to_number` support converting between values of string and

Reviewer 2: ```suggestion
Functions such as `to_number` and `to_char` support converting between values of string and
```

Author: Done.

Decimal type. Such functions accept format strings indicating how to map between these types.

### Syntax

Number format strings support the following syntax:

```
  { ' [ S ] [ $ ] 
      [ 0 | 9 | G | , ] [...] 
      [ . | D ] 
      [ 0 | 9 ] [...] 
      [ $ ] [ PR | MI | S ] ' }
```

### Elements

Each number format string can contain the following elements (case insensitive):

- **`0`** or **`9`**

  Specifies an expected digit between `0` and `9`.

  A sequence of 0 or 9 in the format string matches a sequence of digits in the input string. If the
  0/9 sequence starts with 0 and is before the decimal point, it can only match a digit sequence of
  the same size. Otherwise, if the sequence starts with 9 or is after the decimal point, it can

>   match a digit sequence that has the same or smaller size.

Reviewer 2: Since this doc is for both parsing and formatting, let's rephrase it a little bit:
```
A sequence of 0 or 9 in the format string matches a sequence of digits with the same or smaller size.
If the 0/9 sequence starts with 0 and is before the decimal point, it requires matching the number of digits:
when parsing only a digit sequence of the same size can be matched, when formatting the
digit sequence will be left-padded with zeros in the result string to reach the same size.
Otherwise, the 0/9 sequence can match any digit sequence with the same or smaller size
when parsing, and pad the digit sequence with spaces in the result string when formatting.
Note that, the digit sequence will become a '#' sequence if the size is larger than the 0/9
sequence.
```

Reviewer 2: Let's rephrase other element docs as well. also cc @srielau

Author: Sounds good, done.

  limitations under the License.
---

### Description

Functions such as `to_number` and `try_to_number` support converting between values of string and
Decimal type. Such functions accept format strings indicating how to map between these types.

### Syntax

Number format strings support the following syntax:

```

>   { ' [ S ] [ L | $ ] 

Reviewer 2: do we end up supporting `L`?

Author: Good catch, no, we did not. Removed this part from the syntax.

      [ 0 | 9 | G | , ] [...] 
      [ . | D ] 
      [ 0 | 9 ] [...] 
      [ L | $ ] [ PR | MI | S ] ' }
```

### Elements

Each number format string can contain the following elements (case insensitive):

- **`0`** or **`9`**

  Specifies an expected digit between `0` and `9`.

>   A `0` to the left of the decimal point indicates that at least this many digits must be present.

Reviewer 2: shall we use the doc in the functions?
```
           A sequence of 0 or 9 in the format
           string matches a sequence of digits in the input string. If the 0/9 sequence starts with
           0 and is before the decimal point, it can only match a digit sequence of the same size.
           Otherwise, if the sequence starts with 9 or is after the decimal point, it can match a
           digit sequence that has the same or smaller size.
```

Author: Sure, sounds good. Updated this to the docs here.

  limitations under the License.
---

### Description

Functions such as `to_number` and `to_char` support converting between values of string and
Decimal type. Such functions accept format strings indicating how to map between these types.

### Syntax

Number format strings support the following syntax:

```

>   { ' [ S ] [ $ ] 

Reviewer 2: ```suggestion
  { ' [ S | MI ] [ $ ] 
```

      [ 0 | 9 | G | , ] [...] 
      [ . | D ] 
      [ 0 | 9 ] [...] 
      [ $ ] [ PR | MI | S ] ' }
```

### Elements

Each number format string can contain the following elements (case insensitive):

- **`0`** or **`9`**

  Specifies an expected digit between `0` and `9`.

  A sequence of 0 or 9 in the format string matches a sequence of digits with the same or smaller
  size. If the 0/9 sequence starts with 0 and is before the decimal point, it requires matching the
  number of digits: when parsing, it matches only a digit sequence of the same size; when
  formatting, the result string adds left-padding with zeros to the digit sequence to reach the
  same size. Otherwise, the 0/9 sequence matches any digit sequence with the same or smaller size
  when parsing, and pads the digit sequence with spaces in the result string when formatting. Note
  that the digit sequence will become a '#' sequence if the size is larger than the 0/9 sequence.

- **`.`** or **`D`**

>   Specifies the position of the decimal point.

Author: Done.

>   The input value does not need to include a decimal point.

Reviewer 2: ```suggestion
  When parsing, the input string does not need to include a decimal point.
```

Author: Done.

- **`,`** or **`G`**

  Specifies the position of the `,` grouping (thousands) separator.

>   There must be a `0` or `9` to the left and right of each grouping separator. 

Reviewer 2: ```suggestion
  There must be a `0` or `9` to the left and right of each grouping separator. When parsing,
  the input string must match the grouping separator relevant for the size of the number.
```

Author: Done.

- **`$`**

  Specifies the location of the `$` currency sign. This character may only be specified once.

- **`S`** 

  Specifies the position of an optional '+' or '-' sign. This character may only be specified once.

- **`MI`**

>   Specifies an optional `-` sign at the end, but no `+`.

Reviewer 2: Seems our implementation allows `MI` at the beginning. Is it intentional and we should fix the doc, or it's a mistake?

Author: Good catch -- we ended up allowing `MI` at the beginning as well during conversations while building the feature. I updated these docs accordingly.

Reviewer 2: did you push any updates?

Author: sorry about that -- I forgot to push. Updated now :)

> SELECT to_number('-$12,345.67', 'S$999,099.99');
  -12345.67
 
-- The '$' sign is not optional.
> SELECT to_number('5', '$9');
  Error: the input string does not match the given number format
 
-- The plus sign is optional, and so are fractional digits.
> SELECT to_number('$345', 'S$999,099.99');
  345.00
 
-- The format requires at least three digits.
> SELECT to_number('$45', 'S$999,099.99');
  Error: the input string does not match the given number format
 
-- The format requires at least three digits.
> SELECT to_number('$045', 'S$999,099.99');
  45.00
 
-- MI indicates an optional minus sign at the beginning or end of the input string.
> SELECT to_number('1234-', '999999MI');
  -1234
 
-- PR indicates optional wrapping angel brakets.
> SELECT to_number('9', '999PR')
  9
```

#### The `try_to_number` function:

```sql
-- The '$' sign is not optional.
> SELECT try_to_number('5', '$9');
  NULL
 
-- The format requires at least three digits.
> SELECT try_to_number('$45', 'S$999,099.99');
  NULL
```

#### The `to_char` function:

```sql
> SELECT to_char(decimal(454), '999');
  "454"

-- '99' can format digit sequence with a smaller size.
> SELECT to_char(decimal(1), '99.9');

>   "1.0"

Reviewer 2: ```suggestion
  " 1.0"
```
